File: Kiro/java/src/main/java/gridneighborhoods/Position.java
package gridneighborhoods;

import java.util.Objects;

/**
 * Represents a position in a 2D grid with row and column coordinates.
 * 
 * The coordinate system uses (0,0) as the bottom-left corner of the grid.
 */
public class Position {
    private final int row;
    private final int column;
    
    /**
     * Initialize a position with row and column coordinates.
     * 
     * @param row The row coordinate (>= 0)
     * @param column The column coordinate (>= 0)
     * @throws IllegalArgumentException If row or column is negative
     */
    public Position(int row, int column) {
        if (row < 0) {
            throw new IllegalArgumentException("Row must be non-negative, got " + row);
        }
        if (column < 0) {
            throw new IllegalArgumentException("Column must be non-negative, got " + column);
        }
        this.row = row;
        this.column = column;
    }
    
    /**
     * Calculate Manhattan distance to another position.
     * 
     * @param other Another Position instance
     * @return The Manhattan distance as a non-negative integer
     */
    public int manhattanDistance(Position other) {
        return Math.abs(this.row - other.row) + Math.abs(this.column - other.column);
    }
    
    public int getRow() {
        return row;
    }
    
    public int getColumn() {
        return column;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Position position = (Position) o;
        return row == position.row && column == position.column;
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(row, column);
    }
    
    @Override
    public String toString() {
        return "Position(" + row + ", " + column + ")";
    }
}

File: Kiro/ruby/lib/grid_neighborhoods/position.rb
# frozen_string_literal: true

module GridNeighborhoods
  # A cell position in the grid, identified by (row, column).
  # (0,0) represents the bottom-left corner per the spec glossary.
  #
  # Requirements 2.1-2.3: Manhattan distance calculation.
  class Position
    attr_reader :row, :column

    def initialize(row, column)
      @row = row
      @column = column
    end

    # Manhattan distance to another position: |r1-r2| + |c1-c2|.
    # Requirement 2.1: computes |row1 - row2| + |column1 - column2|.
    # Requirement 2.2: always non-negative.
    # Requirement 2.3: 0 when positions are identical.
    def manhattan_distance(other)
      (row - other.row).abs + (column - other.column).abs
    end

    # Value equality so positions can live in a Set / Hash key.
    def ==(other)
      other.is_a?(Position) && row == other.row && column == other.column
    end
    alias eql? ==

    def hash
      [row, column].hash
    end

    def to_s
      "(#{row},#{column})"
    end
  end
end
